pretyping/reductionops.ml

```diff
@@ -1751,29 +1751,43 @@ let is_sort env sigma t =
 let whd_betaiota_deltazeta_for_iota_state ts env sigma s =
   let refold = false in
   let tactic_mode = false in
-  let all = CClosure.RedFlags.red_add_transparent CClosure.all ts in
+  let all' = CClosure.RedFlags.red_add_transparent CClosure.all ts in
   (* Unset the sharing flag to get a call-by-name reduction. This matters for
      the shape of the generated term. *)
   let env' = Environ.set_typing_flags { (Environ.typing_flags env) with Declarations.share_reduction = false } env in
-  let whd c =
-    let r = clos_whd_flags all env' sigma (Stack.zip sigma c) in
-    let (r, args) = Termops.decompose_app_vect sigma r in
-    r, Stack.append_app args []
+  let whd_opt c =
+    let open CClosure in
+    let evars ev = safe_evar_value sigma ev in
+    let infos = create_clos_infos ~evars all' env' in
+    let tab = create_tab () in
```

Comment:
Could we share the tab across calls since sharing is unset?

Reply:
No, I am willing to be this would lead to trouble. There are probably invariants that will be messed up, so better safe than sorry.